### ISSUE TITLE
rpma_cq_get_completion: failed rpma_completion(op_context=0x0, op_status=local protection error)

### ISSUE THREAD

**Issue by zhijianli88** [pmem/rpma#1142]
QEMU supports baking its vNVDIMM memory-backend-file to a fsdax file(/mnt/pmem1/nv-128m.img). Howver qemu failed to do migration where it would use RDMA transfer data from source(Host A) fsdax to destination(Host) fsdax.

Then i go to here, and find examples/04-write-to-persistent, but it also failed when i specify <pmem-path> to a fsdax file in both server and client side.

Server and Client are the same host.
```
[root@iaas-rpma 04-write-to-persistent]# lspci | grep ConnectX-4
01:00.0 Ethernet controller: Mellanox Technologies MT27710 Family [ConnectX-4 Lx]
01:00.1 Ethernet controller: Mellanox Technologies MT27710 Family [ConnectX-4 Lx]

[root@iaas-rpma 04-write-to-persistent]# ibv_devinfo -d rocep1s0f0
hca_id:	rocep1s0f0
	transport:			InfiniBand (0)
	fw_ver:				14.25.1020
	node_guid:			b859:9f03:00ea:bfea
	sys_image_guid:			b859:9f03:00ea:bfea
	vendor_id:			0x02c9
	vendor_part_id:			4117
	hw_ver:				0x0
	board_id:			MT_2420110034
	phys_port_cnt:			1
		port:	1
			state:			PORT_ACTIVE (4)
			max_mtu:		4096 (5)
			active_mtu:		1024 (3)
			sm_lid:			0
			port_lid:		0
			port_lmc:		0x00
			link_layer:		Ethernet


[root@iaas-rpma 04-write-to-persistent]# mount | grep pmem
/dev/pmem0 on /mnt/pmem0 type ext4 (rw,relatime,seclabel,dax=always)
/dev/pmem1 on /mnt/pmem1 type ext4 (rw,relatime,seclabel,dax=always)
```

Server:
```
# ./server 192.168.1.11 8888 /mnt/pmem0/nv-128m.img
```

Client:
```
[root@iaas-rpma 04-write-to-persistent]# ./client 192.168.1.11 8888 /mnt/pmem1/nv-128m.img 
Next value: ¡Hola Mundo!
Jul 15 13:08:28.935490 [51866] *NOTE*  conn_req.c: 317: rpma_conn_req_new: Requesting a connection to 192.168.1.11:8888
Jul 15 13:08:28.938581 [51866] *NOTE*  conn.c: 189: rpma_conn_next_event: Connection established
Jul 15 13:08:28.939324 [51866] *INFO*  mr.c: 417: rpma_mr_remote_from_descriptor: new rpma_mr_remote(raddr=0x7f6200800000, size=134217728, rkey=0x1a797, usage=0x9)
Jul 15 13:08:28.943267 [51866] *WARN*  cq.c: 160: rpma_cq_get_completion: failed rpma_completion(op_context=0x0, op_status=local protection error)
unexpected cmpl.op value (3 != 0)
Jul 15 13:08:28.944281 [51866] *NOTE*  conn.c: 229: rpma_conn_disconnect: Requesting for disconnection
Jul 15 13:08:28.945402 [51866] *NOTE*  conn.c: 189: rpma_conn_next_event: Connection closed

```

So i wonder if this use case has been already supported ?

**Comment by janekmi** [comment]
Hi @zhijianli88. Thank you for very the very detailed report.

Since you are using a single machine and have only a single RDMA port active, I assume you also have a single IP address for both client and server?
Can you confirm this configuration of yours works for DRAM-based traffic?

**Comment by zhijianli88** [comment]
> Hi @zhijianli88. Thank you for very the very detailed report.
> 
> Since you are using a single machine and have only a single RDMA port active, 

>I assume you also have a single IP address for both client and server?
Yes, that's true

> Can you confirm this configuration of yours works for DRAM-based traffic?

I setup another host for cient, but it still fails. Do I miss something ?
```
[root@iaas-rdma 04-write-to-persistent]# ./client 192.168.1.11 8888 /mnt/pmem0/nv-128m.img
Next value: 
Jul 15 15:11:06.039138 [1353] *NOTE*  conn_req.c: 369: rpma_conn_req_new: Requesting a connection to 192.168.1.11:8888
Jul 15 15:11:06.048227 [1353] *NOTE*  conn.c: 190: rpma_conn_next_event: Connection established
Jul 15 15:11:06.050113 [1353] *INFO*  mr.c: 417: rpma_mr_remote_from_descriptor: new rpma_mr_remote(raddr=0x7f22f2200000, size=134217728, rkey=0xb8a8, usage=0x9)
Jul 15 15:11:06.056360 [1353] *WARN*  conn.c: 598: rpma_conn_completion_get: failed rpma_completion(op_context=0x0, op_status=local protection error)
unexpected cmpl.op value (3 != 0)
Jul 15 15:11:06.056803 [1353] *NOTE*  conn.c: 230: rpma_conn_disconnect: Requesting for disconnection
Jul 15 15:11:06.058483 [1353] *NOTE*  conn.c: 190: rpma_conn_next_event: Connection closed
[root@iaas-rdma 04-write-to-persistent]# 
[root@iaas-rdma 04-write-to-persistent]# ethtool eth1
Settings for eth1:
	Supported ports: [ FIBRE ]
	Supported link modes:   1000baseKX/Full
	                        10000baseKR/Full
	                        25000baseCR/Full
	                        25000baseKR/Full
	                        25000baseSR/Full
	Supported pause frame use: Symmetric
	Supports auto-negotiation: Yes
	Supported FEC modes: None	 RS	 BASER
	Advertised link modes:  1000baseKX/Full
	                        10000baseKR/Full
	                        25000baseCR/Full
	                        25000baseKR/Full
	                        25000baseSR/Full
	Advertised pause frame use: Symmetric
	Advertised auto-negotiation: Yes
	Advertised FEC modes: None	 RS	 BASER
	Link partner advertised link modes:  Not reported
	Link partner advertised pause frame use: No
	Link partner advertised auto-negotiation: Yes
	Link partner advertised FEC modes: Not reported
	Speed: 25000Mb/s
	Duplex: Full
	Auto-negotiation: on
	Port: FIBRE
	PHYAD: 0
	Transceiver: internal
	Supports Wake-on: d
	Wake-on: d
        Current message level: 0x00000004 (4)
                               link
	Link detected: yes
[root@iaas-rdma 04-write-to-persistent]# ifconfig eth1
eth1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.1.10  netmask 255.255.255.0  broadcast 192.168.1.255
        inet6 fe80::ba59:9fff:feea:bff6  prefixlen 64  scopeid 0x20<link>
        ether b8:59:9f:ea:bf:f6  txqueuelen 1000  (Ethernet)
        RX packets 5  bytes 338 (338.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 21  bytes 1584 (1.5 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

[root@iaas-rdma 04-write-to-persistent]# mount | grep pmem
/dev/pmem0 on /mnt/pmem0 type ext4 (rw,relatime,dax=always)

```

**Comment by janekmi** [comment]
Still, please try without PMem just in case you have an issue with fundamentals.

**Comment by zhijianli88** [comment]
> Still, please try without PMem just in case you have an issue with fundamentals.

Well, devdax and default are okay

without Pmem:
```
[root@iaas-rdma 04-write-to-persistent]# ./client 192.168.1.11 8888 
Next value: Hello world!
Jul 15 17:44:31.367594 [143631] *NOTE*  conn_req.c: 369: rpma_conn_req_new: Requesting a connection to 192.168.1.11:8888
Jul 15 17:44:31.372058 [143631] *NOTE*  conn.c: 190: rpma_conn_next_event: Connection established
Jul 15 17:44:31.372117 [143631] *INFO*  mr.c: 417: rpma_mr_remote_from_descriptor: new rpma_mr_remote(raddr=0xf3d000, size=1024, rkey=0x11202, usage=0x9)
translating...
Translation: ¡Hola Mundo!
Jul 15 17:44:31.372685 [143631] *NOTE*  conn.c: 230: rpma_conn_disconnect: Requesting for disconnection
Jul 15 17:44:31.373716 [143631] *NOTE*  conn.c: 190: rpma_conn_next_event: Connection closed

[root@iaas-rpma 04-write-to-persistent]# ./server 192.168.1.11 8888 
Jul 15 17:49:47.044624 [2720] *NOTE*  ep.c:  92: rpma_ep_listen: Waiting for incoming connection on 192.168.1.11:8888
Jul 15 17:49:49.334719 [2720] *NOTE*  conn.c: 189: rpma_conn_next_event: Connection established
Jul 15 17:49:49.335340 [2720] *NOTE*  conn.c: 189: rpma_conn_next_event: Connection closed
Jul 15 17:49:49.336132 [2720] *NOTE*  conn.c: 229: rpma_conn_disconnect: Requesting for disconnection
New value: Hello world!

```

With devdax pmem
```
[root@iaas-rdma 04-write-to-persistent]# ./client 192.168.1.11 8888 /dev/dax1.0 
Next value: ¡Hola Mundo!
Jul 15 17:46:26.925849 [143639] *NOTE*  conn_req.c: 369: rpma_conn_req_new: Requesting a connection to 192.168.1.11:8888
Jul 15 17:46:26.930659 [143639] *NOTE*  conn.c: 190: rpma_conn_next_event: Connection established
Jul 15 17:46:26.940834 [143639] *INFO*  mr.c: 417: rpma_mr_remote_from_descriptor: new rpma_mr_remote(raddr=0x7f9d00000000, size=526385152, rkey=0x2a1a, usage=0x9)
translating...
Translation: Hello world!
Jul 15 17:46:26.951230 [143639] *NOTE*  conn.c: 230: rpma_conn_disconnect: Requesting for disconnection
Jul 15 17:46:26.952535 [143639] *NOTE*  conn.c: 190: rpma_conn_next_event: Connection closed

[root@iaas-rpma 04-write-to-persistent]# ./server 192.168.1.11 8888 /dev/dax1.0 
Old value: Hello world!
Jul 15 17:51:39.535470 [2725] *NOTE*  ep.c:  92: rpma_ep_listen: Waiting for incoming connection on 192.168.1.11:8888
Jul 15 17:51:44.890901 [2725] *NOTE*  conn.c: 189: rpma_conn_next_event: Connection established
Jul 15 17:51:44.911617 [2725] *NOTE*  conn.c: 189: rpma_conn_next_event: Connection closed
Jul 15 17:51:44.912432 [2725] *NOTE*  conn.c: 229: rpma_conn_disconnect: Requesting for disconnection
New value: ¡Hola Mundo!

```

**Comment by janekmi** [comment]
> Well, devdax and default are okay

Please give me also a server output for the FileSystemDAX case.

**Comment by zhijianli88** [comment]
the server side exits with 0

```
[root@iaas-rpma 04-write-to-persistent]# while ./server 192.168.1.11 8888 /mnt/pmem0/nv-128m.img ; do :; done
Jul 15 15:35:10.970777 [52841] *NOTE*  ep.c:  92: rpma_ep_listen: Waiting for incoming connection on 192.168.1.11:8888
Jul 15 15:35:12.429764 [52841] *NOTE*  conn.c: 189: rpma_conn_next_event: Connection established
Jul 15 15:35:12.435821 [52841] *NOTE*  conn.c: 189: rpma_conn_next_event: Connection closed
Jul 15 15:35:12.436218 [52841] *NOTE*  conn.c: 229: rpma_conn_disconnect: Requesting for disconnection
New value: 
Jul 15 15:35:12.464258 [52842] *NOTE*  ep.c:  92: rpma_ep_listen: Waiting for incoming connection on 192.168.1.11:8888
Jul 15 15:35:16.901160 [52842] *NOTE*  conn.c: 189: rpma_conn_next_event: Connection established
Jul 15 15:35:16.905835 [52842] *NOTE*  conn.c: 189: rpma_conn_next_event: Connection closed
Jul 15 15:35:16.906109 [52842] *NOTE*  conn.c: 229: rpma_conn_disconnect: Requesting for disconnection
New value: 
Jul 15 15:35:16.935017 [52844] *NOTE*  ep.c:  92: rpma_ep_listen: Waiting for incoming connection on 192.168.1.11:8888
Jul 15 15:35:17.800357 [52844] *NOTE*  conn.c: 189: rpma_conn_next_event: Connection established
Jul 15 15:35:17.805018 [52844] *NOTE*  conn.c: 189: rpma_conn_next_event: Connection closed
Jul 15 15:35:17.805392 [52844] *NOTE*  conn.c: 229: rpma_conn_disconnect: Requesting for disconnection
New value: 
Jul 15 15:35:17.834709 [52845] *NOTE*  ep.c:  92: rpma_ep_listen: Waiting for incoming connection on 192.168.1.11:8888
Jul 15 15:35:18.762185 [52845] *NOTE*  conn.c: 189: rpma_conn_next_event: Connection established
Jul 15 15:35:18.767174 [52845] *NOTE*  conn.c: 189: rpma_conn_next_event: Connection closed
Jul 15 15:35:18.767429 [52845] *NOTE*  conn.c: 229: rpma_conn_disconnect: Requesting for disconnection
New value: 
```

**Comment by zhijianli88** [comment]
I ran this test with kernel v5.11 and v5.12

**Comment by zhijianli88** [comment]
ping

**Comment by janekmi** [comment]
We are working on the FileSystemDAX issue. I have to ask you for patience.

**Comment by janekmi** [comment]
The issue does not reproduce without QEMU using directly PMEM+RDMA. We are using a slightly older kernel 4.18.0-193.28.1.el8_2.x86_64 but I would not expect kernel downgrade will solve the issue. I believe the QEMU developers should take a look at this issue.

Especially the error message indicates something really weird has happened:
`unexpected cmpl.op value (3 != 0)`

Which translates to `RPMA_OP_SEND != RPMA_OP_READ` where this example does not use `rpma_send()`. On the libibverbs level, it suggests you got a completion for `IBV_WC_SEND` without posting this kind of RDMA operation.

**Comment by zhijianli88** [comment]
> The issue does not reproduce without QEMU using directly PMEM+RDMA. We are using a slightly older kernel 4.18.0-193.28.1.el8_2.x86_64 but I would not expect kernel downgrade will solve the issue. I believe the QEMU developers should take a look at this issue.

emm, i'm QEMU a developer :), The qemu currently cannot fully support the fsdax qemu migration, so  I'm working on fsdax mem-backend-file migration. Let me try the older kernel first.


> 
> Especially the error message indicates something really weird has happened:
> `unexpected cmpl.op value (3 != 0)`
> 
> Which translates to `RPMA_OP_SEND != RPMA_OP_READ` where this example does not use `rpma_send()`. On the libibverbs level, it suggests you got a completion for `IBV_WC_SEND` without posting this kind of RDMA operation.

thank you for the explanation。

**Comment by yangx-jy** [comment]
> > Especially the error message indicates something really weird has happened:
> > `unexpected cmpl.op value (3 != 0)`
> > Which translates to `RPMA_OP_SEND != RPMA_OP_READ` where this example does not use `rpma_send()`. On the libibverbs level, it suggests you got a completion for `IBV_WC_SEND` without posting this kind of RDMA operation.
> 
> thank you for the explanation。

Hi @zhijianli88 @janekmi 

When  `wc.status` is not IBV_WC_SUCCESS, `wc.opcode` is not set and is still the initialized value (i.e. IBV_WC_SEND == 0) so that `cmpl.op` is set to RPMA_OP_SEND (3) by IBV_WC_SEND (0).  In this case, `wc.opcode` and `cmpl.op` make no sense.
Ref: #1170

**Comment by yangx-jy** [comment]
Hi @zhijianli88 @janekmi 

It seems a kernel bug.  I have confirmed that v5.10 and newer kernel can hit the error (In other word, v5.9 and older kernel is OK).

**Comment by zhijianli88** [comment]
> Hi @zhijianli88 @janekmi
> 
> It seems a kernel bug. I have confirmed that v5.10 and newer kernel can hit the error (In other word, v5.9 and older kernel is OK).

Great, as you said it worked  in my local 5.8
Linux iaas-rpma 5.8.15-301.fc33.x86_64 #1 SMP Thu Oct 15 16:58:06 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

**Comment by zhijianli88** [comment]
[lizhijian@yl linux]$ git bisect log
git bisect start
# good: [bbf5c979011a099af5dc76498918ed7df445635b] Linux 5.9
git bisect good bbf5c979011a099af5dc76498918ed7df445635b
# bad: [2c85ebc57b3e1817b6ce1a6b703928e113a90442] Linux 5.10
git bisect bad 2c85ebc57b3e1817b6ce1a6b703928e113a90442
# good: [4d0e9df5e43dba52d38b251e3b909df8fa1110be] lib, uaccess: add failure injection to usercopy functions
git bisect good 4d0e9df5e43dba52d38b251e3b909df8fa1110be
# bad: [6694875ef8045cdb1e6712ee9b68fe08763507d8] ext4: indicate that fast_commit is available via /sys/fs/ext4/feature/...
git bisect bad 6694875ef8045cdb1e6712ee9b68fe08763507d8
# good: [14c914fcb515c424177bb6848cc2858ebfe717a8] Merge tag 'wireless-drivers-next-2020-10-02' of git://git.kernel.org/pub/scm/linux/kernel/git/kvalo/wireless-drivers-next
git bisect good 14c914fcb515c424177bb6848cc2858ebfe717a8
# good: [6f78b9acf04fbf9ede7f4265e7282f9fb39d2c8c] Merge tag 'mtd/for-5.10' of git://git.kernel.org/pub/scm/linux/kernel/git/mtd/linux
git bisect good 6f78b9acf04fbf9ede7f4265e7282f9fb39d2c8c
# bad: [bbe85027ce8019c73ab99ad1c2603e2dcd1afa49] Merge tag 'xfs-5.10-merge-5' of git://git.kernel.org/pub/scm/fs/xfs/xfs-linux
git bisect bad bbe85027ce8019c73ab99ad1c2603e2dcd1afa49
# bad: [9d9af1007bc08971953ae915d88dc9bb21344b53] Merge tag 'perf-tools-for-v5.10-2020-10-15' of git://git.kernel.org/pub/scm/linux/kernel/git/acme/linux
git bisect bad 9d9af1007bc08971953ae915d88dc9bb21344b53
# good: [21c2fe94abb2abe894e6aabe6b4e84a255c8d339] RDMA/mthca: Combine special QP struct with mthca QP
git bisect good 21c2fe94abb2abe894e6aabe6b4e84a255c8d339
# good: [dbaa1b3d9afba3c050d365245a36616ae3f425a7] Merge branch 'perf/urgent' into perf/core
git bisect good dbaa1b3d9afba3c050d365245a36616ae3f425a7
# bad: [c7a198c700763ac89abbb166378f546aeb9afb33] RDMA/ucma: Fix use after free in destroy id flow
git bisect bad c7a198c700763ac89abbb166378f546aeb9afb33
# bad: [5ce2dced8e95e76ff7439863a118a053a7fc6f91] RDMA/ipoib: Set rtnl_link_ops for ipoib interfaces
git bisect bad 5ce2dced8e95e76ff7439863a118a053a7fc6f91
# bad: [a03bfc37d59de316436c46f5691c5a972ed57c82] RDMA/mlx5: Sync device with CPU pages upon ODP MR registration
git bisect bad a03bfc37d59de316436c46f5691c5a972ed57c82
# good: [a6f0b08dbaf289c3c57284e16ac8043140f2139b] RDMA/core: Remove ucontext->closing
git bisect good a6f0b08dbaf289c3c57284e16ac8043140f2139b
# bad: [36f30e486dce22345c2dd3a3ba439c12cd67f6ba] IB/core: Improve ODP to use hmm_range_fault()
git bisect bad 36f30e486dce22345c2dd3a3ba439c12cd67f6ba
# good: [2ee9bf346fbfd1dad0933b9eb3a4c2c0979b633e] RDMA/addr: Fix race with netevent_callback()/rdma_addr_cancel()
git bisect good 2ee9bf346fbfd1dad0933b9eb3a4c2c0979b633e
# first bad commit: [36f30e486dce22345c2dd3a3ba439c12cd67f6ba] IB/core: Improve ODP to use hmm_range_fault()

Note: some commit have to apply a extra patch to avoid a kernel panic.
> git cherry-pick d4c5da5 # dax: Fix stack overflow when mounting fsdax pmem device

**Comment by janekmi** [comment]
Sorry for lagging on this topic. But it seems you have figured out everything by yourself.

Thank you @yangx-jy for stepping in!

@zhijianli88 do you plan to submit a kernel bug?

**Comment by zhijianli88** [comment]
> Sorry for lagging on this topic. But it seems you have figured out everything by yourself.
> 
> Thank you @yangx-jy for stepping in!
> 
> @zhijianli88 do you plan to submit a kernel bug?

I did that, but have not got significant response from community yet. 
https://lkml.org/lkml/2021/8/4/124

**Comment by zhijianli88** [comment]
patch have been queued into mm tree
https://www.spinics.net/lists/stable/msg494441.html

**Comment by zhijianli88** [comment]
Fixed by https://github.com/torvalds/linux/commit/4b42fb213678d2b6a9eeea92a9be200f23e49583

**Comment by zhijianli88** [comment]
this patch  was shipped by [v5.15-rc1](https://github.com/torvalds/linux/commit/6880fa6c56601bb8ed59df6c30fd390cc5f6dd8f)
